File: 0097-interleaving-string/0097-interleaving-string.py
class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        
        if len(s1) + len(s2) != len(s3):
            return False
        
        cache = {(len(s1), len(s2)): True}
        
        def dfs(i, j):
            if (i, j) in cache:
                return cache[(i, j)]
            k = i + j
            if j < len(s2) and s3[k] == s2[j]:
                cache[(i, j)] = dfs(i, j + 1)
                if cache[(i, j)]:
                    return True
            if i < len(s1) and s3[k] == s1[i]:
                cache[(i, j)] = dfs(i + 1, j)
                return cache[(i, j)]
            
            cache[(i, j)] = False
            return cache[(i, j)]
        
        return dfs(0, 0)
                
        
        
        
        

File: 0097-interleaving-string/0097-interleaving-string.cpp
class Solution {
public:
    string s1;
    string s2;
    string s3;
    vector<vector<int>> dp;

    bool helper(int i, int j) {
        if (i == s1.size() and j == s2.size()) return true;
        if(dp[i][j] != - 1) return dp[i][j];
        int k = i + j;
        bool ans = false;
        if (i < s1.size() and s1[i] == s3[k]) {
            ans = ans or helper(i + 1, j);
        }

        if (j < s2.size() and s2[j] == s3[k]) {
            ans = ans or helper(i, j + 1);
        }

        return dp[i][j] = ans;
    }

    bool isInterleave(string s1, string s2, string s3) {
        this->s1 = s1;
        this->s2 = s2;
        this->s3 = s3;
        if (s1.size() + s2.size() != s3.size()) return false;
        dp.resize(s1.size() + 1, vector<int>(s2.size() + 1, -1));
        return helper(0, 0); 
    }
};